### DISCUSSION TITLE
Is it possible to support OpenAI's latest o1-preview and o1-mini series models?

### DISCUSSION THREAD

**Discussion by wangzhe0912** [open-webui/open-webui#5390] [General]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by Fusseldieb** [comment]
Most of us don't even have it yet, so we can't build a pipeline for it.

**Comment by tkafka** [comment]
Well, some do, and I expect the API to be the same, so it would seem it's just about adding model names, right? And we can iterate on that later ...

**Comment by m8dhouse** [comment]
For the moment the new models are in beta an only available via API for **Tier 5** users.

**Comment by mjbommar** [comment]
There shouldn't be any real issue with the API, but it's currently being filtered out thanks to the non-sense naming scheme.

Just add .startswith("o1") as a condition here and it's fine.

https://github.com/open-webui/open-webui/blob/e510c8b11f0f898796635cf09aa93a3d72c4c2ee/backend/open_webui/apps/openai/main.py#L223

**Reply by Legalcomplex** [reply]
o1 is only available for Tier 4-5 user 
![image](https://github.com/user-attachments/assets/64b13495-1bc5-421c-a32f-0c79d48bcbb8)

**Reply by mjbommar** [reply]
Yup, but if you do have it, this change is all you need to do.  I have it running locally in webui or directly via the normal /messages chat endpoint:
```
>>> import alea_llm_client
>>> model = alea_llm_client.OpenAIModel(model="o1-preview")
>>> model.chat("how many r's are there in strawberry")
ModelResponse(choices=['The word **"strawberry"** contains **three** letter \'r\'s. Here\'s a breakdown of each letter in the word to illustrate where the \'r\'s appear:\n\n1. **S**\n2. **T**\n3. **R** ← first \'r\'\n4. **A**\n5. **W**\n6. **B**\n7. **E**\n8. **R** ← second \'r\'\n9. **R** ← third \'r\'\n10. **Y**\n\nSo, the letter \'r\' appears in the 3rd, 8th, and 9th positions of the word "strawberry," totaling three occurrences.'], metadata={'model': 'o1-preview', 'usage': {'prompt_tokens': 16, 'completion_tokens': 596, 'total_tokens': 612, 'completion_tokens_details': {'reasoning_tokens': 448}}}, text='The word **"strawberry"** contains **three** letter \'r\'s. Here\'s a breakdown of each letter in the word to illustrate where the \'r\'s appear:\n\n1. **S**\n2. **T**\n3. **R** ← first \'r\'\n4. **A**\n5. **W**\n6. **B**\n7. **E**\n8. **R** ← second \'r\'\n9. **R** ← third \'r\'\n10. **Y**\n\nSo, the letter \'r\' appears in the 3rd, 8th, and 9th positions of the word "strawberry," totaling three occurrences.')
```

**Reply by mjbommar** [reply]
Pushed to branch here if you want to cherrypick or build from it:
https://github.com/alea-institute/open-webui/tree/support-openai-o1-strawberry-models

**Reply by johnlindquist** [reply]
@mjbommar so I've confirmed my key works with o1-mini through curl (from a tier 5 plan). It also works using other SDKs. But when I try it with the latest webui, it can see the model but it 400 errors out... I'm brand new to webui and would love any help. Any ideas?

![CleanShot 2024-09-13 at 15 58 12](https://github.com/user-attachments/assets/47348e0e-d2aa-4c2b-857e-b5d841273c49)


Just showing it works on other models
![CleanShot 2024-09-13 at 15 57 24](https://github.com/user-attachments/assets/594b6ded-bd59-40d8-96bb-695edc444237)

**Reply by mjbommar** [reply]
@johnlindquist did you build from `main` or `dev`?  it seems like @tjbck has been trying a few different things but i'm not sure if this repo has a working version yet

**Reply by johnlindquist** [reply]
@mjbommar I've tried both. Just saw @tjbck 's pr as merged/closed so I assumed working. So your fork is working?

**Reply by cheahjs** [reply]
The `o1` models do not currently support streaming, and there's currently no way to disable streaming for chats.

**Reply by johnlindquist** [reply]
Ah, that makes sense, thanks @cheahjs

**Reply by mjbommar** [reply]
@cheahjs @johnlindquist the issue is in svelte, not the backend though.  there is an issue with max_tokens->max_completion_tokens, but you can still set `stream: False` too.  i will try to push a complete version built off current `dev`

**Comment by mjbommar** [comment]
@cheahjs  @johnlindquist 

![image](https://github.com/user-attachments/assets/c4b2facb-2c6e-4da3-82df-fa26fc08eaaa)

https://github.com/alea-institute/open-webui/tree/o1-support-fake-streaming-response

with latest dev branch, it was easier to get full frontend support working by faking the streaming response at this point and patch the headers.  

i'm not sure what's going on with streaming support in the frontend anymore, so maybe you just want to remove the original backend flow for non-streaming models.  (this pattern should work for any non-streaming model, regardless of whether it's an OAI model)

there's still also a question of whether you want to default to a separate OAI model for the chat title.  i didn't deal with that, but suggestion would be to let the user config an alternative/mapping.

anyway, this is at least enough for most people to get started

**Reply by mjbommar** [reply]
k, i cleaned this up and PR'd.  this is somewhat ugly, but unless the project wants to adopt a proper abstraction layer (which would also help for azure, anthropic, bedrock, vllm, etc...), then there's going to be a lot of kludge.

https://github.com/open-webui/open-webui/pull/5421

**Comment by spammenotinoz** [comment]
If it helps anyone, I just consume o1 using Open-WebUI from OpenRouter (happen to be using LiteLLM).  Worked from Day1, OpenRouter must be streaming the response somehow as this worked before LiteLLM supported o1.  Probably don't need LiteLLM but I use it for reporting and other features.
